nat46/modules/nat46-module.c

```diff
@@ -86,13 +100,58 @@ static char *get_devname(char **ptail)
 	return devname;
 }
 
+static void
+nat46_proc_cmd(char *cmd)
+{
+	char *arg = NULL;
+	char *devname = NULL;
+
+	while(NULL != (arg = get_next_arg (&cmd))) {
+		if(0 == strcmp(arg, "add")) {
+			devname = get_devname(&cmd);
+			printk(KERN_INFO "nat46: adding device (%s)\n", devname);
+			nat46_create(devname);
+		}
+		else if(0 == strcmp(arg, "del")) {
+			devname = get_devname(&cmd);
+			printk(KERN_INFO "nat46: deleting device (%s)\n", devname);
+			nat46_destroy(devname);
+		}
+		else if(0 == strcmp(arg, "config")) {
+			devname = get_devname(&cmd);
+			printk(KERN_INFO "nat46: configure device (%s) with '%s'\n", devname, cmd);
+			nat46_configure(devname, cmd);
+		}
+		else if(0 == strcmp(arg, "insert")) {
+			devname = get_devname(&cmd);
+			printk(KERN_INFO "nat46: insert new rule into device (%s) with '%s'\n",
+				devname, cmd);
+			nat46_insert(devname, cmd);
+		}
+	}
+}
+
+#ifdef PROTO_NETLINK
+static void
+nat46_nl_recv_msg(struct sk_buff *skb_in)
+{
+	struct nlmsghdr *nl_hdr;
+	int pid;
+	char *cmd = NULL;
+
+	nl_hdr = (struct nlmsghdr *) skb_in->data;
+	pid = nl_hdr->nlmsg_pid;
+	cmd = (char *) nlmsg_data (nl_hdr);
+
+	nat46_proc_cmd(cmd);
```

Comment:
nat46_proc_cmd assumes a zero-terminated string which it is always the case for the procfs case. Here the code takes an arbitrary pointer from the message and passes it to the string parsing function without validation. You have to make sure here that the user can trigger the parsing code go beyond the boundary of the cmd.

Also, the whole point of using the netlink and similar structures is to have the parsing of the data done in the user space, and supply well-defined data structures to the kernel space. So - could you describe a little, what is the rationale to use the netlink to pass the strings vs. just using procfs ?

Reply:
Our system uses Netlink as the interface between kernel and user space, so I thought of having a Netlink socket support for this module to keep things consistent.  Thanks for your feedback. I'll keep your suggestions in mind as I update my patch.